### PR TITLE
Bring back OpenTelemetry

### PR DESCRIPTION
**Pull request by pbrisbin**
This branch can exist and be used by apps that are looking to experiment
with `s/X-Ray/OTel + Datadog/`. Once we're doing that in our main apps,
we will merge and release it.

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1204160835912882

### CODE REVIEW

File: library/Freckle/App/OpenTelemetry.hs [hunk 1]
```diff
@@ -1,19 +1,67 @@
...
+--   tracerL = lens appTracer $ \x y -> x { appTracer = y }
+--
+-- loadApp f = do
+--   -- ...
+--   withTracerProvider $ \tracerProvider -> do
+--     let appTracer = makeTracer tracerProvider "my-app" tracerOptions
+--     f App {..}
+-- @
+--
+-- You may need to do even if you don't plan to manual trace things, in order to
```
Comment (pbrisbin) [inline library/Freckle/App/OpenTelemetry.hs:19]:
```suggestion
-- You may need to do this even if you don't plan to manually trace things, in order to
```

File: library/Network/Wai/Middleware/OpenTelemetry.hs [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    Just (traceId, _) -> do
+      let
+        traceIdInt = ddTraceId traceId
+        traceIdIntBS = BS8.pack $ show traceIdInt
+
+      withThreadContext ["trace_id" .= traceIdInt] $ addHeaders
+        [("X-Datadog-Trace-Id", traceIdIntBS)]
+        app
+        request
+        respond
```
Comment (pbrisbin) [inline library/Network/Wai/Middleware/OpenTelemetry.hs:49]: NOTE: everything seems to function without pulling in the actual (unreleased) [Datadog propagator](https://github.com/iand675/hs-opentelemetry/blob/main/propagators/datadog/src/OpenTelemetry/Propagator/Datadog.hs), which would do similar stuff with the headers and trace-ids to make them Datadog compatible. All we're really doing here is ensuring the `trace_id` present in the logs is DD-formatted so it correlates with the traces in Datadog.

File: tests/Network/Wai/Middleware/OpenTelemetrySpec.hs [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+import Test.Hspec
+
+spec :: Spec
+spec = do
+  describe "ddTraceId" $ do
+    it "converts correctly to Word64" $ do
+      let
+        Right traceId =
+          baseEncodedToTraceId Base16 "4e2ef0fa8943f417f6209f2b6f6ddcea"
+      ddTraceId traceId `shouldBe` 17735350341486894314
```
Comment (pbrisbin) [inline tests/Network/Wai/Middleware/OpenTelemetrySpec.hs:20]: This example here was taken from a concrete Trace in Dev, comparing the hex version to the i64 one.

**Comment by pbrisbin** [comment]
I don't plan to merge this yet without more testing. But it is ready to review.

File: tests/Network/Wai/Middleware/OpenTelemetrySpec.hs (hunk 3 above)
Comment (StevenXL) [reply tests/Network/Wai/Middleware/OpenTelemetrySpec.hs:20]: I wonder if some version of the comment would be useful in the code itself.

**Review by StevenXL — APPROVED** [review]

File: library/Freckle/App.hs [hunk 4]
```diff
@@ -114,13 +119,16 @@
 -- >   appConfig{..} <- loadConfig
 -- >   appLogger <- newLogger configLoggerSettings
 -- >   appSqlPool <- runLoggerLoggingT appLogger $ makePostgresPool configDbPoolSize
--- >   withStatsClient configStatsSettings $ \appStatsClient -> do
--- >     f App{..}
+-- >   withTracerProvider $ \tracerProvider -> do
```
[the remark this answers is not in the document]
Comment (pbrisbin) [reply library/Freckle/App.hs:122]: The variables are prefixed `OTEL_` but yes, exactly.

File: tests/Network/Wai/Middleware/OpenTelemetrySpec.hs (hunk 3 above)
Comment (pbrisbin) [reply tests/Network/Wai/Middleware/OpenTelemetrySpec.hs:20]: Will do.

Comment (pbrisbin) [inline tests/Network/Wai/Middleware/OpenTelemetrySpec.hs:20]:
```suggestion
      -- Example taken from a real test trace
      ddTraceId traceId `shouldBe` 17735350341486894314
```

**Comment by pbrisbin** [comment]
Going to open fresh.

**Pull request closed without merging**